sklearn/partial_dependence.py

```diff
@@ -0,0 +1,589 @@
+
+    features = np.asarray(features, dtype=np.int32, order='C').ravel()
+    if any(not (0 <= f < n_features) for f in features):
+        raise ValueError('all features must be in [0, %d]'
+                         % (n_features - 1))
+
+    grid, values = _grid_from_X(X[:, features], percentiles,
+                                grid_resolution)
+    if method == 'brute':
+        averaged_predictions = _partial_dependence_brute(est, grid,
+                                                         features, X)
+    else:
+        averaged_predictions = _partial_dependence_recursion(est, grid,
+                                                             features)
+
+    # reshape averaged_predictions to
+    # (n_outputs, n_values_feature_0, # n_values_feature_1, ...)
+    averaged_predictions = averaged_predictions.reshape(
+        -1, *[val.shape[0] for val in values])
+
+    return averaged_predictions, values
+
+
+def plot_partial_dependence(est, X, features, feature_names=None,
+                            target=None, n_cols=3, grid_resolution=100,
+                            percentiles=(0.05, 0.95), method='auto',
+                            n_jobs=1, verbose=0, fig=None, line_kw=None,
+                            contour_kw=None, **fig_kw):
+    """Partial dependence plots.
+
+    The ``len(features)`` plots are arranged in a grid with ``n_cols``
+    columns. Two-way partial dependence plots are plotted as contour plots.
+
+    Read more in the :ref:`User Guide <partial_dependence>`.
+
+    Parameters
+    ----------
+    est : BaseEstimator
+        A fitted classification or regression model. Classifiers must have a
+        ``predict_proba()`` method. Multioutput-multiclass estimators aren't
+        supported.
+    X : array-like, shape=(n_samples, n_features)
+        The data to use to build the grid of values on which the dependence
+        will be evaluated. This is usually the training data.
+    features : list of ints or strings, or tuples of ints or strings
+        The target features for which to create the PDPs.
+        If features[i] is an int or a string, a one-way PDP is created; if
+        features[i] is a tuple, a two-way PDP is created. Each tuple must be
+        of size 2.
+        if any entry is a string, then it must be in ``feature_names``.
+    feature_names : seq of str, shape=(n_features,)
+        Name of each feature; feature_names[i] holds the name of the feature
+        with index i.
+    target : int, optional (default=None)
+        - In a multiclass setting, specifies the class for which the PDPs
+          should be computed. Note that for binary classification, the
+          positive class (index 1) is always used.
+        - In a multioutput setting, specifies the task for which the PDPs
+          should be computed
+        Ignored in binary classification or classical regression settings.
+    n_cols : int, optional (default=3)
+        The number of columns in the grid plot.
```

Comment:
Should this be reduced if `len(features) < 3`?

Reply:
It is ;)

```
    n_cols = min(n_cols, len(features))
```